proposals/0021-vk-coop-matrix.md

```diff
@@ -0,0 +1,346 @@
+
+```c++
+
+// The base cooperative matrix class. The template arguments correspond to the
+// operands in the OpTypeCooperativeMatrixKHR instruction.
+template <typename ComponentType, Scope scope, uint rows, uint columns,
+          CooperativeMatrixUse use>
+class CooperativeMatrix {
+  template <class NewComponentType>
+  CooperativeMatrix<NewComponentType, scope, rows, columns, use> cast();
+
+  // Apply OpSNegate or OFNegate, depending on ComponentType, in a element by
+  // element manner.
+  CooperativeMatrix negate();
+
+  // Apply OpIAdd or OFAdd, depending on ComponentType, in a element by element
+  // manner.
+  CooperativeMatrix operator+(CooperativeMatrix other);
+
+  // Apply OpISub or OFSub, depending on ComponentType, in a element by element
+  // manner.
+  CooperativeMatrix operator-(CooperativeMatrix other);
+
+  // Apply OpIMul or OFMul, depending on ComponentType, in a element by element
+  // manner.
+  CooperativeMatrix operator*(CooperativeMatrix other);
+
+  // Apply OpSDiv, OpUDiv or OFDiv, depending on ComponentType, in a element by
+  // element manner.
+  CooperativeMatrix operator/(CooperativeMatrix other);
+
+  // Apply OpMatrixTimesScalar in a element by element manner.
+  CooperativeMatrix operator*(ComponentType scalar);
+
+  // Note: The SPIR-V specification allows the load and store function to access
+  // workgroup data (groupshared in HLSL). This call does not accept group
+  // shared data because HLSL cannot pass the groupshared array by reference. It
+  // is impossible to implement a function that will load and store from group
+  // sharded data. This can be revisited when references are added to HLSL.
+
+  // Store the cooperative matrix using OpCooperativeMatrixStoreKHR to data[i]
+  // using memory layout RowMajorKHR.￼
+  template <class Type>
+  void StoreRowMajor(RWStructuredBuffer<Type> data, uint32_t index);
+
+  // Store the cooperative matrix using OpCooperativeMatrixStoreKHR to data[i]
+  // using memory layout ColumnMajorKHR.￼
+  template <class Type>
+  void StoreColumnMajor(RWStructuredBuffer<Type> data, uint32_t index);
+
+  // Store the cooperative matrix using OpCooperativeMatrixStoreKHR to data[i]
+  // using memory layout RowMajorKHR and the given stride and memory access
+  // mask.
+  template <class Type>
+  void StoreRowMajor(RWStructuredBuffer<Type> data, uint32_t index,
+                     uint32_t stride, MemoryAccessMask memoryAccessMask);
+
+  // Store the cooperative matrix using OpCooperativeMatrixStoreKHR to data[i]
+  // using memory layout ColumnMajorKHR and the given stride and memory access
+  // mask.
+  template <class Type>
+  void StoreColumnMajor(RWStructuredBuffer<Type> data, uint32_t index,
+                        uint32_t stride, MemoryAccessMask memoryAccessMask);
+
+  // Constructs a cooperative matrix with all values initialized to v. Note that
+  // all active threads must have the same value for v.
+  static CooperativeMatrix splat(ComponentType v);
```

Comment:
Can't add constructors. :shrug:

Reply:
I _think_ we'll probably add constructors in HLSL 202y. They'd be pretty hard to add in DXC for a lot of not-great reasons. My initial PR to propose them is [here](https://github.com/microsoft/hlsl-specs/pull/5). GitHub closed it because the original fork doesn't exist anymore.